// uix-mat-number-spinner/esm2015/lib/mat-number-spinner.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ChangeDetectorRef } from "@angular/core";
export class MatNumberSpinnerComponent {
    /**
     * @param {?} cdr
     */
    constructor(cdr) {
        this.cdr = cdr;
        this.classSpinner = "";
        this.valueChange = new EventEmitter();
        this.increase = 1;
        this.decrease = 1;
        this.fix = 1;
        this.add = (a, b) => a + b;
        this.substract = (a, b) => a - b;
        this.parse = (...args) => args.map(value => this.handleParse(value));
        this.operation = (f, g) => (...args) => this.handleParse(g(...f(...args)));
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.cdr.detectChanges();
    }
    /**
     * @return {?}
     */
    up() {
        const /** @type {?} */ result = this.operation(this.parse, this.add);
        this.value = result(this.value, this.increase);
        this.valueChange.emit(this.value);
    }
    /**
     * @return {?}
     */
    down() {
        const /** @type {?} */ result = this.operation(this.parse, this.substract);
        this.value = result(this.value, this.decrease);
        this.valueChange.emit(this.value);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    handleParse(value) {
        return +parseFloat(value).toFixed(this.fix);
    }
}
MatNumberSpinnerComponent.decorators = [
    { type: Component, args: [{
                selector: "uix-mat-number-spinner",
                template: `
  <div class="uix-spinner-arrows">
    <div [ngClass]="['uix-spinner-'+classSpinner]">
      <i class="uix-spinner-up" (click)="up()">
        <div class="uix-spinner-arrow uix-spinner-arrow-up"></div>
      </i>
      <i class="uix-spinner-down" (click)="down()">
        <div class="uix-spinner-arrow uix-spinner-arrow-down"></div>
      </i>
    </div>
  </div>
  `,
                styles: [`::ng-deep uix-mat-number-spinner+input[type=number]::-webkit-inner-spin-button,input[type=number]::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}.uix-spinner-outline .uix-spinner-up{top:18px}.uix-spinner-outline .uix-spinner-down{top:47px}.uix-spinner-arrows{position:absolute;width:20px;left:auto;right:-10px;top:-23px;z-index:999999999999}.uix-spinner-arrow{border:solid #636363;border-width:0 2px 2px 0;display:inline-block;padding:3px}.uix-spinner-arrow-up{transform:rotate(-135deg);-webkit-transform:rotate(-135deg)}.uix-spinner-arrow-down{transform:rotate(45deg);-webkit-transform:rotate(45deg)}.uix-spinner-up{top:5px;width:13px;position:absolute;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;-o-user-select:none;user-select:none}.uix-spinner-up:hover .uix-spinner-arrow{border:solid #73c0ff;border-width:0 2px 2px 0}.uix-spinner-down{top:30px;width:13px;position:absolute;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;-o-user-select:none;user-select:none}.uix-spinner-down:hover .uix-spinner-arrow{border:solid #73c0ff;border-width:0 2px 2px 0}`]
            },] },
];
/** @nocollapse */
MatNumberSpinnerComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
MatNumberSpinnerComponent.propDecorators = {
    classSpinner: [{ type: Input, args: ["type",] }],
    value: [{ type: Input }],
    valueChange: [{ type: Output }],
    increase: [{ type: Input }],
    decrease: [{ type: Input }],
    fix: [{ type: Input }]
};
function MatNumberSpinnerComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.classSpinner;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.value;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.valueChange;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.increase;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.decrease;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.fix;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.add;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.substract;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.parse;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.operation;
    /** @type {?} */
    MatNumberSpinnerComponent.prototype.cdr;
}
