// uix-mat-number-spinner/fesm5/uix-mat-number-spinner.js
import { __spread } from 'tslib';
import { Component, Input, Output, EventEmitter, ChangeDetectorRef, NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var MatNumberSpinnerComponent = /** @class */ (function () {
    function MatNumberSpinnerComponent(cdr) {
        var _this = this;
        this.cdr = cdr;
        this.classSpinner = "";
        this.valueChange = new EventEmitter();
        this.increase = 1;
        this.decrease = 1;
        this.fix = 1;
        this.add = function (a, b) { return a + b; };
        this.substract = function (a, b) { return a - b; };
        this.parse = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return args.map(function (value) { return _this.handleParse(value); });
        };
        this.operation = function (f, g) { return function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return _this.handleParse(g.apply(void 0, __spread(f.apply(void 0, __spread(args)))));
        }; };
    }
    /**
     * @return {?}
     */
    MatNumberSpinnerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.cdr.detectChanges();
    };
    /**
     * @return {?}
     */
    MatNumberSpinnerComponent.prototype.up = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ result = this.operation(this.parse, this.add);
        this.value = result(this.value, this.increase);
        this.valueChange.emit(this.value);
    };
    /**
     * @return {?}
     */
    MatNumberSpinnerComponent.prototype.down = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ result = this.operation(this.parse, this.substract);
        this.value = result(this.value, this.decrease);
        this.valueChange.emit(this.value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    MatNumberSpinnerComponent.prototype.handleParse = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return +parseFloat(value).toFixed(this.fix);
    };
    MatNumberSpinnerComponent.decorators = [
        { type: Component, args: [{
                    selector: "uix-mat-number-spinner",
                    template: "\n  <div class=\"uix-spinner-arrows\">\n    <div [ngClass]=\"['uix-spinner-'+classSpinner]\">\n      <i class=\"uix-spinner-up\" (click)=\"up()\">\n        <div class=\"uix-spinner-arrow uix-spinner-arrow-up\"></div>\n      </i>\n      <i class=\"uix-spinner-down\" (click)=\"down()\">\n        <div class=\"uix-spinner-arrow uix-spinner-arrow-down\"></div>\n      </i>\n    </div>\n  </div>\n  ",
                    styles: ["::ng-deep uix-mat-number-spinner+input[type=number]::-webkit-inner-spin-button,input[type=number]::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}.uix-spinner-outline .uix-spinner-up{top:18px}.uix-spinner-outline .uix-spinner-down{top:47px}.uix-spinner-arrows{position:absolute;width:20px;left:auto;right:-10px;top:-23px;z-index:999999999999}.uix-spinner-arrow{border:solid #636363;border-width:0 2px 2px 0;display:inline-block;padding:3px}.uix-spinner-arrow-up{transform:rotate(-135deg);-webkit-transform:rotate(-135deg)}.uix-spinner-arrow-down{transform:rotate(45deg);-webkit-transform:rotate(45deg)}.uix-spinner-up{top:5px;width:13px;position:absolute;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;-o-user-select:none;user-select:none}.uix-spinner-up:hover .uix-spinner-arrow{border:solid #73c0ff;border-width:0 2px 2px 0}.uix-spinner-down{top:30px;width:13px;position:absolute;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;-o-user-select:none;user-select:none}.uix-spinner-down:hover .uix-spinner-arrow{border:solid #73c0ff;border-width:0 2px 2px 0}"]
                },] },
    ];
    /** @nocollapse */
    MatNumberSpinnerComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    MatNumberSpinnerComponent.propDecorators = {
        classSpinner: [{ type: Input, args: ["type",] }],
        value: [{ type: Input }],
        valueChange: [{ type: Output }],
        increase: [{ type: Input }],
        decrease: [{ type: Input }],
        fix: [{ type: Input }]
    };
    return MatNumberSpinnerComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var UixMatNumberSpinnerModule = /** @class */ (function () {
    function UixMatNumberSpinnerModule() {
    }
    UixMatNumberSpinnerModule.decorators = [
        { type: NgModule, args: [{
                    imports: [BrowserModule],
                    declarations: [MatNumberSpinnerComponent],
                    exports: [MatNumberSpinnerComponent]
                },] },
    ];
    return UixMatNumberSpinnerModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { MatNumberSpinnerComponent, UixMatNumberSpinnerModule };
