// uix-mat-number-spinner/bundles/uix-mat-number-spinner.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/platform-browser')) :
    typeof define === 'function' && define.amd ? define('uix-mat-number-spinner', ['exports', '@angular/core', '@angular/platform-browser'], factory) :
    (factory((global['uix-mat-number-spinner'] = {}),global.ng.core,global.ng.platformBrowser));
}(this, (function (exports,core,platformBrowser) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var MatNumberSpinnerComponent = (function () {
        function MatNumberSpinnerComponent(cdr) {
            var _this = this;
            this.cdr = cdr;
            this.classSpinner = "";
            this.valueChange = new core.EventEmitter();
            this.increase = 1;
            this.decrease = 1;
            this.fix = 1;
            this.add = function (a, b) { return a + b; };
            this.substract = function (a, b) { return a - b; };
            this.parse = function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                return args.map(function (value) { return _this.handleParse(value); });
            };
            this.operation = function (f, g) {
                return function () {
                    var args = [];
                    for (var _i = 0; _i < arguments.length; _i++) {
                        args[_i] = arguments[_i];
                    }
                    return _this.handleParse(g.apply(void 0, __spread(f.apply(void 0, __spread(args)))));
                };
            };
        }
        /**
         * @return {?}
         */
        MatNumberSpinnerComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.cdr.detectChanges();
            };
        /**
         * @return {?}
         */
        MatNumberSpinnerComponent.prototype.up = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ result = this.operation(this.parse, this.add);
                this.value = result(this.value, this.increase);
                this.valueChange.emit(this.value);
            };
        /**
         * @return {?}
         */
        MatNumberSpinnerComponent.prototype.down = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ result = this.operation(this.parse, this.substract);
                this.value = result(this.value, this.decrease);
                this.valueChange.emit(this.value);
            };
        /**
         * @param {?} value
         * @return {?}
         */
        MatNumberSpinnerComponent.prototype.handleParse = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                return +parseFloat(value).toFixed(this.fix);
            };
        MatNumberSpinnerComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "uix-mat-number-spinner",
                        template: "\n  <div class=\"uix-spinner-arrows\">\n    <div [ngClass]=\"['uix-spinner-'+classSpinner]\">\n      <i class=\"uix-spinner-up\" (click)=\"up()\">\n        <div class=\"uix-spinner-arrow uix-spinner-arrow-up\"></div>\n      </i>\n      <i class=\"uix-spinner-down\" (click)=\"down()\">\n        <div class=\"uix-spinner-arrow uix-spinner-arrow-down\"></div>\n      </i>\n    </div>\n  </div>\n  ",
                        styles: ["::ng-deep uix-mat-number-spinner+input[type=number]::-webkit-inner-spin-button,input[type=number]::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}.uix-spinner-outline .uix-spinner-up{top:18px}.uix-spinner-outline .uix-spinner-down{top:47px}.uix-spinner-arrows{position:absolute;width:20px;left:auto;right:-10px;top:-23px;z-index:999999999999}.uix-spinner-arrow{border:solid #636363;border-width:0 2px 2px 0;display:inline-block;padding:3px}.uix-spinner-arrow-up{transform:rotate(-135deg);-webkit-transform:rotate(-135deg)}.uix-spinner-arrow-down{transform:rotate(45deg);-webkit-transform:rotate(45deg)}.uix-spinner-up{top:5px;width:13px;position:absolute;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;-o-user-select:none;user-select:none}.uix-spinner-up:hover .uix-spinner-arrow{border:solid #73c0ff;border-width:0 2px 2px 0}.uix-spinner-down{top:30px;width:13px;position:absolute;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;-o-user-select:none;user-select:none}.uix-spinner-down:hover .uix-spinner-arrow{border:solid #73c0ff;border-width:0 2px 2px 0}"]
                    },] },
        ];
        /** @nocollapse */
        MatNumberSpinnerComponent.ctorParameters = function () {
            return [
                { type: core.ChangeDetectorRef }
            ];
        };
        MatNumberSpinnerComponent.propDecorators = {
            classSpinner: [{ type: core.Input, args: ["type",] }],
            value: [{ type: core.Input }],
            valueChange: [{ type: core.Output }],
            increase: [{ type: core.Input }],
            decrease: [{ type: core.Input }],
            fix: [{ type: core.Input }]
        };
        return MatNumberSpinnerComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var UixMatNumberSpinnerModule = (function () {
        function UixMatNumberSpinnerModule() {
        }
        UixMatNumberSpinnerModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [platformBrowser.BrowserModule],
                        declarations: [MatNumberSpinnerComponent],
                        exports: [MatNumberSpinnerComponent]
                    },] },
        ];
        return UixMatNumberSpinnerModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.MatNumberSpinnerComponent = MatNumberSpinnerComponent;
    exports.UixMatNumberSpinnerModule = UixMatNumberSpinnerModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
